test/unit/org/apache/cassandra/cql3/GuardrailPagingTest.java

```diff
@@ -0,0 +1,125 @@
+        createTable("CREATE TABLE IF NOT EXISTS %s (k INT, c INT, v TEXT, PRIMARY KEY(k, c))");
+
+        for (int i = 0; i < partitionCount; i++)
+            for (int j = 0; j < rowsPerPartition; j++)
+                execute("INSERT INTO %s (k, c, v) VALUES (?, ?, ?)", i, j, "test message");
+    }
+
+
+    private ResultMessage.Rows selectWithPaging(int size, PageSize.PageUnit unit) throws InvalidRequestException
+    {
+        PageSize pageSize = new PageSize(size, unit);
+        QueryOptions options = QueryOptions.create(ConsistencyLevel.LOCAL_QUORUM,
+                                                   Collections.emptyList(),
+                                                   false,
+                                                   pageSize,
+                                                   null,
+                                                   ConsistencyLevel.LOCAL_SERIAL,
+                                                   ProtocolVersion.CURRENT,
+                                                   KEYSPACE);
+
+        ClientState clientState = ClientState.forExternalCalls(AuthenticatedUser.ANONYMOUS_USER);
+        clientState.setKeyspace(KEYSPACE);
+        QueryState queryState = new QueryState(clientState);
+
+        QueryHandler.Prepared prepared = QueryProcessor.prepareInternal(String.format("SELECT * FROM %s.%s", KEYSPACE, currentTable()));
+        return (ResultMessage.Rows) prepared.statement.execute(queryState, options, System.nanoTime());
+    }
+
+    /**
+     * Test that the number of returned rows per page is silently limited to fit into the guardrail hard limit
+     */
+    @Test
+    public void testQueryWithPagedByRows() throws Throwable
+    {
+        // ask for more rows per page than can fit with the current guardrail
+        int rowLimit = partitionCount * rowsPerPartition;
+        ResultMessage.Rows result = selectWithPaging(rowLimit, PageSize.PageUnit.ROWS);
+        assertTrue(result.result.rows.size() < rowLimit);
+    }
+
+    /**
+     * Test that a query throws with page size that is bigger than the guardrail hard limit
+     */
+    @Test(expected = InvalidRequestException.class)
+    public void testQueryWithLargeBytePagesThrows() throws Throwable
+    {
+        selectWithPaging(140 * 1024, PageSize.PageUnit.BYTES);
+    }
+
+    /**
+     * Test that a query does not throw with page size that is smaller than the guardrail hard limit
+     */
+    @Test
+    public void testQueryWithSmallBytePagesWorks() throws Throwable
+    {
+        int maxPageSize = 100 * 128;
+        ResultMessage.Rows result = selectWithPaging(maxPageSize, PageSize.PageUnit.BYTES);
+        // technically incorrect as we compare a size of encoded message to be sent to a client to the page size,
+        // but we can't know the page at this point.
+        assertTrue(ResultMessage.codec.encodedSize(result, ProtocolVersion.CURRENT) < maxPageSize);
```

Comment:
I was curious about the "technically incorrect" comment and tried to use something like this to get a more accurate count
```
        Integer totalSize = rows.stream().flatMap(byteBuffers -> byteBuffers.stream()).map(byteBuffer -> byteBuffer.remaining())
                                .reduce(0, (subtotal, remaining) -> subtotal + remaining);
```
It works ok, but the value was far less than the `maxPageSize` bytes and I couldn't figure out why.

With the maxPageSize of 12800, it returned 242 rows with the exiting "encodedSize" result of 7830.  My counting byte buffer method got only 4840. Oh well. 🤷🏻‍♂️

Reply:
So, page size is not what you get as a result - it is the approximate amount of memory used by the query results on the coordinator. That data include for example write time and TTL for each single regular column in each row. There is no difference how we do that calculation comparing to DSE, though I don't claim it is correct - as you can see in the ticket, I've created a follow up to include partition key in the calculation as its size is currently skipped